### PR TITLE
init: deprecate SMP init level

### PR DESCRIPTION
**Pull request by gmarull**
The level meaning has never been clear, ie, how it is related to SMP. It
was the last level to be initialized when enabled. Only a couple of
cases were found upstream. They have been moved to one level below
(APPLICATION) to minimize any potential issues.

Users of SMP level should likely look carefully into their code and use
other levels such as POST_KERNEL or APPLICATION.

This PR also introduces `Z_INIT_CHECK_DEPRECATED_LEVEL` so that other
levels can be easily deprecated in the future.

---

Example output if used:

```
[61/134] Building C object CMakeFiles/app.dir/src/main.c.obj
/home/gmarull/ws/nordic/zephyrproject/zephyr/samples/hello_world/src/main.c:22:13: warning: SMP level is deprecated
   22 | SYS_INIT(cb, SMP, 0);
      |             ^~~~~~~~~  
```

### CODE REVIEW

**Review by fabiobaltieri — DISMISSED** [review]

**Review by nashif — DISMISSED** [review]
thought it would be cleaner to just introduce LATE at the end , given that whatever was done in SMP is not really application related, but that works as well.

**Comment by nashif** [comment]
@andyross  if I remember correctly, you introduced this init level, so please give your thoughts.

**Comment by gmarull** [comment]
note: investigating CI issues (deprecation macro), they seem to happen for gcc 11.x, but not on gcc 12/13.

**Comment by gmarull** [comment]
deprecation failures, tested this with https://godbolt.org/

```
#include <stdio.h>

/* for COND_CODE_1 */
#define __DEBRACKET(...) __VA_ARGS__

#define __GET_ARG2_DEBRACKET(ignore_this, val, ...) __DEBRACKET val

#define __COND_CODE(one_or_two_args, _if_code, _else_code) \
	__GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)

#define _XXXX1 _YYYY,

#define Z_COND_CODE_1(_flag, _if_1_code, _else_code) \
	__COND_CODE(_XXXX##_flag, _if_1_code, _else_code)

#define COND_CODE_1(_flag, _if_1_code, _else_code) \
	Z_COND_CODE_1(_flag, _if_1_code, _else_code)

/* for __WARN */

#define __WARN(msg) __WARN1(GCC warning msg)
#define __WARN1(s) _Pragma(#s)

/* for Z_INIT_CHECK_DEPRECATED_LEVEL */
#define Z_INIT_LEVEL_SMP_SMP 1

#define Z_INIT_CHECK_DEPRECATED_LEVEL(level) \
    COND_CODE_1(Z_INIT_LEVEL_SMP_##level, (__WARN("SMP is deprecated")), ())

/* main */
int main(void)
{
    /* should emit a warning */
    Z_INIT_CHECK_DEPRECATED_LEVEL(SMP)
    /* should not emit a warning <--- FAILS for gcc < 12 */
    Z_INIT_CHECK_DEPRECATED_LEVEL(OTHER)
}
```

File: include/zephyr/init.h [hunk 1]
```diff
@@ -135,13 +137,21 @@ struct init_entry {
 	__attribute__((__section__(                                           \
 		".z_init_" #level STRINGIFY(prio)"_" STRINGIFY(sub_prio)"_")))
 
+/**
+ * @brief Issue a warning if the given init level is deprecated.
+ *
+ * @param level Init level
+ */
+#define Z_INIT_CHECK_DEPRECATED_LEVEL(level)                                   \
+	COND_CODE_1(Z_INIT_SMP_##level, (__WARN("SMP level is deprecated")), ())
```
Comment (57300) [inline include/zephyr/init.h:146]: Suggestion for GCC 11. Would this work?

```c
#define Z_INIT_DEPRECATED_EARLY
#define Z_INIT_DEPRECATED_PRE_KERNEL_1
#define Z_INIT_DEPRECATED_PRE_KERNEL_2
#define Z_INIT_DEPRECATED_POST_KERNEL
#define Z_INIT_DEPRECATED_APPLICATION
#define Z_INIT_DEPRECATED_SMP		__WARN("SMP level is deprecated")
```
```suggestion
#define Z_INIT_CHECK_DEPRECATED_LEVEL(level) Z_INIT_DEPRECATED_##level
```

Comment (gmarull) [reply include/zephyr/init.h]: great idea, seems to work! thanks, that saved a lot of investigation time :-)

**Review by 57300 — DISMISSED** [review]

**Review by fabiobaltieri — DISMISSED** [review]

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Using the `APPLICATION` level for in-tree drivers seems like a hack. I'd rather see us go with introducing a new level for "late" initialization, that does not hook into what is intended for the application to use.

**Comment by gmarull** [comment]
> Using the `APPLICATION` level for in-tree drivers seems like a hack. I'd rather see us go with introducing a new level for "late" initialization, that does not hook into what is intended for the application to use.

I'd honestly not introduce more levels without clarifying first what SMP meant. Maybe that code could actually run on POST_KERNEL, but who knows. APPLICATION level is already a late level, so if adding LATE, I'd drop APPLICATION (APPLICATION feels like "just before main/app code", so late).

**Comment by fabiobaltieri** [comment]
> I'd honestly not introduce more levels without clarifying first what SMP meant. Maybe that code could actually run on POST_KERNEL, but who knows. APPLICATION level is already a late level, so if adding LATE, I'd drop APPLICATION (APPLICATION feels like "just before main/app code", so late).

I think APPLICATION should be left to the application, although some functionalities implemented in the core repository are arguably standalone application features (mcumgr...). Regardless it seems clear these levels should be revisited at a later stage, would not add more now.

**Comment by henrikbrixandersen** [comment]
> I think APPLICATION should be left to the application, although some functionalities implemented in the core repository are arguably standalone application features (mcumgr...). Regardless it seems clear these levels should be revisited at a later stage, would not add more now.

If we're against adding more or renaming until we revisit, why even touch the SMP one now?

**Comment by gmarull** [comment]
> > I think APPLICATION should be left to the application, although some functionalities implemented in the core repository are arguably standalone application features (mcumgr...). Regardless it seems clear these levels should be revisited at a later stage, would not add more now.
> 
> If we're against adding more or renaming until we revisit, why even touch the SMP one now?

I think removing a level known not to be useful helps to make things simpler, adding new ones does not. I understood this pre-cleanup was welcome to have a better picture.

**Comment by gmarull** [comment]
> Using the `APPLICATION` level for in-tree drivers seems like a hack. I'd rather see us go with introducing a new level for "late" initialization, that does not hook into what is intended for the application to use.

also note, any code using SYS_INIT is **NOT** a driver. A driver, strictly speaking, must be a `struct device`. We have a few exceptions like system timer (they could all be singleton devices really) and interrupt controllers (which is a bit of a mess in Zephyr).

**Comment by henrikbrixandersen** [comment]
> also note, any code using SYS_INIT is **NOT** a driver. A driver, strictly speaking, must be a `struct device`. We have a few exceptions like system timer (they could all be singleton devices really) and interrupt controllers (which is a bit of a mess in Zephyr).

I disagree with the statement that a piece of code is only a "driver" if it uses a `struct device`. This may be desired, but we currently have singleton drivers which use `SYS_INIT()` (you can add pinctrl drivers to your list above).

**Comment by nashif** [comment]
The major value i see in this PR is the ability to deprecate an init level, we are taking `SMP` as the first level to be deprecated, because, nevermind of how it applies and how useful it is to have a dedicated level for `SMP`, the name just does not make sense, so we could keep existing behaviour using a `LATE` hook in the system, outside of `APPLICATION`.

The next step would be the deprecation of `PRE_KERNEL_1` and `PRE_KERNEL_2` and merging them into one (`PRE_KERNEL`), then adding `PLATFORM` (or whatever we want to call it) and moving many of the` soc_init()`calls with priority 0 into this new level.

This will give us a much cleaner canvas to continue with the other work @gmarull  has in the other PR and it will be non-breaking change that we can get in 3.5.

IMO we should just pause for a minute and figure out the overall init level plan, what we want to keep, add and what will be deprecated and work toward that.

**Comment by gmarull** [comment]
> > also note, any code using SYS_INIT is **NOT** a driver. A driver, strictly speaking, must be a `struct device`. We have a few exceptions like system timer (they could all be singleton devices really) and interrupt controllers (which is a bit of a mess in Zephyr).
> 
> I disagree with the statement that a piece of code is only a "driver" if it uses a `struct device`. This may be desired, but we currently have singleton drivers which use `SYS_INIT()` (you can add pinctrl drivers to your list above).

all these cases could (and should) likely be moved to `struct device`.

**Comment by gmarull** [comment]
Added DNM, as I've found this:

https://github.com/zephyrproject-rtos/zephyr/blob/2c2124de76a273c139914569444daa434b63a864/kernel/init.c#L302-L343

So it looks like SMP stuff is initialized after application level. Shouldn't SMP be operational in POST_KERNEL? Or, why it goes after APPLICATION level, for example? @andyross could you shed some light here?

**Comment by nashif** [comment]
hmm, actually it was @dcpleung who introduced SMP level, see 4e1637b54ebe01ff38d62b8b85cf16821527ee38. @dcpleung  can you please provide some background?

**Comment by andyross** [comment]
No strong feelings here.  IIRC the original intent was that this was an initlevel that gets run after the SMP cores are started, so subsystems that have parallelizable loads can put their initialization handling there and get some plausible performance benefit.  But obviously you'd have to spawn threads (or workqueue, or otherwise interact with the OS, etc...) to do that, and there's no reason you can't spawn a thread from other init levels (it just won't run right away).

**Comment by dcpleung** [comment]
It's been a while... but IIRC, the original idea is that some drivers need to be initialized only after other CPUs are up. This was used with some IPC/IPM drivers where early init would somehow crash other processors at power up.

**Comment by gmarull** [comment]
@dcpleung @andyross so my questions are:

- Isn't it safe to start SMP just before POST_KERNEL, that is, code running in POST_KERNEL will have a Kernel with all cores up
- Why is SMP started after APPLICATION?

I feel like this level was added as a quick workaround, but things can probably be improved.

**Comment by dcpleung** [comment]
The SMP init level is supposed to run after `z_smp_init()` where CPUs are being brought up. Commit 2b210cb3db0501867adec4f74044ac557bd81cd9 added the "if boot delay" condition to `z_smp_init()` (and it only) after introduction of the SMP init level, so it is a bit confusing. The SMP init level should also have been enclosed inside the if block.

**Comment by andyross** [comment]
> Isn't it safe to start SMP just before POST_KERNEL, that is, code running in POST_KERNEL will have a Kernel with all cores up

This is a theory vs. practice thing.  Sure, in principle I don't see any reason to separate them; "post kernel" surely means once everything is running, and everything includes extra CPUs.

In practice, what this means is that existing initialization hooks that have worked for many releases will suddenly start racing against the threads they're creating, and probably show up bugs in synchronization that were previously hidden.  Linux, for example, went through a long stability mess trying to parallelize its initialization.   Whether we'll see that or not I don't know -- really Zephyr doesn't have that many SMP platforms still.

**Comment by nashif** [comment]
@gmarull  lets put the deprecation of the SMP init level on hold for now and shift to the items discussed and planned. Probably worth filing a bug about this so we can look into the smp init code seperately and come back to this later. @dcpleung  can you please spend some time on this and see what can be done here and if we can move things from SMP level to some other level or if we need to run those functions that late (after APPLICATION)?

**Comment by dcpleung** [comment]
The `ipm_cavs_idc.c` driver needs to enable IRQ for each core for inter-processor communication only after that core has been up. IIRC, any interrupt would wake up the core to go through the power up process. So we need to be very careful on when we enable interrupt, or else the core(s) would be powering up at the time we are not ready. We cannot assume the hardware state as if it has power-cycled itself to reset state because the audio DSP can be restarted by power cycling only the code (e.g. restarting after crash), while the peripherals (in this case the IPC mechanism) may still be powered and have not gone through reset. Because the audio DSP is part of a SoC which means there may not be a total power cycling of all related hardware. Unless we decouple the interrupt enabling to the SoC layer, we will still need an init level which only starts after `z_smp_init()` has completed. Though, this decoupling would make the driver no longer self-contained as it requires SoC code to make it fully working.

**Comment by dcpleung** [comment]
Just a thought, we could also introduce SMP post initialization (maybe `z_smp_post_init()`) which is implemented on the SoC layer to finish whatever needs to be done after all CPUs are up. Though, it makes the IDC driver no longer self-contained but you can get rid of the SMP init level.

**Comment by gmarull** [comment]
> Just a thought, we could also introduce SMP post initialization (maybe `z_smp_post_init()`) which is implemented on the SoC layer to finish whatever needs to be done after all CPUs are up. Though, it makes the IDC driver no longer self-contained but you can get rid of the SMP init level.

I think this is what would probably make more sense, ie, SMP specific hooks. I see SYS_INIT as a kind of generic, configurable initializer, but as soon as you have something that needs to be called at the right spot, hooks are more predictable and less error-prone. A similar thing happens for SoC init calls, most if not all are set up to run at PRE_KERNEL_1 + hardcoded priority 0 (indicating configurability is not needed). Also, from what I read, we could most likely initialize SMP before entering POST_KERNEL.

**Comment by gmarull** [comment]
> @gmarull lets put the deprecation of the SMP init level on hold for now and shift to the items discussed and planned. Probably worth filing a bug about this so we can look into the smp init code seperately and come back to this later. @dcpleung can you please spend some time on this and see what can be done here and if we can move things from SMP level to some other level or if we need to run those functions that late (after APPLICATION)?

Agree it should not block, but it's good to have this discussion as it will probably lead to a more consistent design (e.g. SoC init hooks? SMP hooks? New levels? Maybe change SMP initialization?)

**Comment by henrikbrixandersen** [comment]
I am all for converting some of these init calls to init hooks (SoC/platform initialization, SMP initialization, perhaps even board initialization?).

**Comment by nashif** [comment]
> I am all for converting some of these init calls to init hooks (SoC/platform initialization, SMP initialization, perhaps even board initialization?).

hook me up

**Comment by nashif** [comment]
> I am all for converting some of these init calls to init hooks (SoC/platform initialization, SMP initialization, perhaps even board initialization?).

@gmarull  @henrikbrixandersen  something like this? https://github.com/zephyrproject-rtos/zephyr/pull/62595

**Comment by fabiobaltieri** [comment]
> The next step would be the deprecation of `PRE_KERNEL_1` and `PRE_KERNEL_2` and merging them into one (`PRE_KERNEL`), then adding `PLATFORM` (or whatever we want to call it) and moving many of the` soc_init()`calls with priority 0 into this new level.

(posting this link once again) https://elixir.bootlin.com/linux/latest/source/include/linux/init.h#L300

Linux has "core" and "arch", maybe one of those two? Then I'm thinking maybe "SUBSYS" and "DEVICE" (or "DRIVER") though we have some subsystem that use the drivers, so are more like services... SUBSYS -> DRIVER -> SERVICE?

Anyway I have https://github.com/zephyrproject-rtos/zephyr/pull/62459 out for producing nice prints of init orders, I did it for troubleshooting other stuff but that may be useful to print some real life application example and see if what combination of level would work well.

**Comment by dcpleung** [comment]
The only in-tree driver using the SMP init level is being removed via https://github.com/zephyrproject-rtos/zephyr/pull/62608.

**Comment by dcpleung** [comment]
Though, another thing to consider is that all INIT levels provide a centralized way of doing various initialization. Having init hooks means we are adding another way of doing things which will cause some confusion.

**Comment by gmarull** [comment]
> Though, another thing to consider is that all INIT levels provide a centralized way of doing various initialization. Having init hooks means we are adding another way of doing things which will cause some confusion.

Well, centralized as a mechanism, but we have a bunch of priorities defined everywhere that can easily enter into conflict. I think that it is much safer to have hooks for things that need to run at a specific point, i.e., no configurability needed.

**Comment by nashif** [comment]
> The only in-tree driver using the SMP init level is being removed via #62608.

we have this one 
drivers/interrupt_controller/intc_arcv2_irq_unit.c:SYS_INIT(arc_shared_intc_update_post_smp, SMP, 0);

**Comment by gmarull** [comment]
> > The only in-tree driver using the SMP init level is being removed via #62608.
> 
> we have this one
> 
> `drivers/interrupt_controller/intc_arcv2_irq_unit.c:SYS_INIT(arc_shared_intc_update_post_smp, SMP, 0);`

@dcpleung is SMP level required for this one? Should we propose an SMP hook if so?

**Comment by nashif** [comment]
> > > The only in-tree driver using the SMP init level is being removed via #62608.
> > 
> > 
> > we have this one
> > `drivers/interrupt_controller/intc_arcv2_irq_unit.c:SYS_INIT(arc_shared_intc_update_post_smp, SMP, 0);`
> 
> @dcpleung is SMP level required for this one? Should we propose an SMP hook if so?

need to check with @evgeniy-paltsev  @ruuddw @abrodkin , was added in https://github.com/zephyrproject-rtos/zephyr/pull/52387

**Comment by ruuddw** [comment]
The main requirement for arc_shared_intc_update_post_smp is to get called after all local, per cpu interrupt controllers have been initialized and set up to process IDU dispatched interrupts (round robbing sent to one of the SMP cpus). All interrupts get processed at cpu 0 until this function is called.

But, what looks a bit suspicious is the requirement (and assert) that this function is called on the primary cpu. A hook may be better for that, and the SMP init level design intent (parallel initialization) seems what is not wanted for this function...

**Pull request closed without merging**